src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -59,6 +59,11 @@ internal WebApplicationBuilder(WebApplicationOptions options, Action<IHostBuilde
 
         bootstrapHostBuilder.ConfigureWebHostDefaults(webHostBuilder =>
         {
+            webHostBuilder.ConfigureServices(services =>
+            {
+                services.AddRouting();
```

Comment:
Are there any concerns (performance or otherwise) with `ConfigureWebHostDefaults` calling `AddRoutingCore()`, and then this line again calling `AddRouting()`? I see most of `AddRoutingCore()` using `TryAdd` for the services. But this takes time (see https://github.com/dotnet/runtime/issues/44728). And I also see `AddRoutingCore()` creates a `new ObservableCollection<EndpointDataSource>();` as well, with some closures being captured.

Reply:
Opening this back up as swapping this out for a call to ```services.Configure<RouteOptions>(...)``` for putting the regex constraint back in created another problem for MVC test cases where we don't want multiple instances of ```IConfigureOptions<RouteOptions>``` registered.

The "hack" to work around this is to create a class that implements ```IPostConfigure<RouteOptions>```  e.g.

```csharp
internal sealed class PostConfigureRouteOptions : IPostConfigureOptions<RouteOptions>
{
    public void PostConfigure(string? name, RouteOptions options)
    {
        // This code is lifted out of AddRouting(...)
        var existingRegexConstraintType = options.TrimmerSafeConstraintMap["regex"];

        // Don't override regex constraint if it has already been overridden
        // this behavior here is just to add it back in if someone calls AddRouting(...)
        // after setting up routing with AddRoutingCore(...).
        if (existingRegexConstraintType == typeof(RegexErrorStubRouteConstraint))
        {
            options.SetParameterPolicy<RegexInlineRouteConstraint>("regex");
        }
    }
}
```

Then injected it into DI using:

```csharp
services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<RouteOptions>, PostConfigureRouteOptions>());
```

If I try to simultaneously satisfy the requirements for the MVC tests, and only call AddRouting(...) under normal usage conditions then I'm going to end up having to make the class above public so I can use it across multiple assemblies. Not to mention the fact that TryAddEnumerable from the DI extensions would then become a dependency of Microsoft.AspNetCore - gross!

Reply:
Is there a reason the MVC tests don't want multiple instances of configuring RouteOptions? The tests may just be asserting that the excepted configuration on the options is registered.

Reply:
There is a teams chat where @davidfowl is asserting that IConfigureOptions<RouteOptions> is only registered once. Its a ```DavidFowl.Assert(...)``` if you will ;)

I've reverted back to calling AddRouting(...) from WebHost for now whilst we figure this out. At least in this state the tests pass and we are dealing with a performance consideration vs. failing tests.